enclave-runner/src/usercalls/mod.rs

```diff
@@ -165,18 +165,30 @@ pub trait SyncStream: 'static + Send + Sync {
 }
 
 
-trait SyncListener: 'static + Send + Sync {
-    fn accept(&self) -> IoResult<(FileDesc, Box<dyn ToString>, Box<dyn ToString>)>;
+/// SyncListener lets an implementation implement a slightly modified form of `std::net::TcpListener::accept`.
+pub trait SyncListener: 'static + Send + Sync {
+    /// The enclave may optionally request the local or peer addresses
+    /// be returned in `local_addr` or `peer_addr`, respectively.
+    /// If `local_addr` and/or `peer_addr` are not `None`, they will point to an empty `String`.
+    /// On success, user-space can fill in the strings as appropriate.
+    ///
+    /// The enclave must not make any security decisions based on the local address received.
+    fn accept(&self, local_addr: Option<&mut String>, peer_addr: Option<&mut String>) -> IoResult<Box<dyn SyncStream>>;
 }
 
 impl SyncListener for TcpListener {
-    fn accept(&self) -> IoResult<(FileDesc, Box<dyn ToString>, Box<dyn ToString>)> {
+    fn accept(&self, local_addr: Option<&mut String>, peer_addr: Option<&mut String>) -> IoResult<Box<dyn SyncStream>> {
         TcpListener::accept(self).map(|(s, peer)| {
-            let local = match s.local_addr() {
-                Ok(local) => Box::new(local) as _,
-                Err(_) => Box::new("error") as _,
-            };
-            (FileDesc::stream(s), local, Box::new(peer) as _)
+            if let Some(local_addr) = local_addr {
+                match self.local_addr() {
+                    Ok(local) => {*local_addr = local.to_string()},
+                    Err(_) => {*local_addr = "error".to_string()},
```

Comment:
Why do we set `"error"` here? It doesn't make sense to me -- it's not an IP address.

Reply:
After further discussion, we were thinking it could make more sense to early return from this function if `self.local_addr()` returns error, or alternatively not set `local_addr` in that case.

Reply:
The ABI [specifies](https://edp.fortanix.com/docs/api/fortanix_sgx_abi/struct.Usercalls.html#networking) that it doesn't have to be an IP address. The ABI also [specifies](https://edp.fortanix.com/docs/api/fortanix_sgx_abi/struct.Usercalls.html#method.accept_stream) that the addresses must be returned. Do you have a suggestion for a better return value? The empty string?

Reply:
The empty string or early returning an error.
See http://man7.org/linux/man-pages/man2/getsockname.2.html

Reply:
I don't think there is case where `self.local_addr()` returns an error but we still want to return successfully.

Reply:
cc @mzohreva

Reply:
Sorry to complain, this is unimportant.